rhombus-lib/rhombus/private/amalgam/error.rhm

```diff
@@ -0,0 +1,164 @@
+#lang rhombus/private/amalgam/core
+import:
+  "core-meta.rkt" open
+  "maybe.rhm" open
+  lib("racket/base.rkt")
+
+use_static
+
+export:
+  error
+
+namespace error:
+  export:
+    Who
+
+    // constructs full message without raising it
+    message
+
+    // builders of main message test
+    annot_msg
+
+    Clause
+
+    // clause builders
+    text
+    val
+    vals
+    annot
+
+    // helper
+    reindent
+
+annot.macro 'Who':
+  'ReadableString || Symbol || Identifier || Operator'
+
+class Clause(msg :: String):
+  nonfinal
+
+// primary `error` function
+// (not backward-comptible with current `error` due to `~who` keyword argument)
+fun error(~srcloc: srcloc :: maybe(Srcloc) = #false,
+          ~who: who :: maybe(error.Who) = #false,
+          ~realm: realm :: Symbol = #'rhombus,
+          // make things like `~exn: Exn.Fail.User` easy; for more complex
+          // exception constructions, use `error.message`, instead, to build
+          // just the message text
+          ~exn: exn :: (Function.of_arity(2)
+                          // also gets srcloc list
+                          || Function.of_arity(3)
+                          // also gets clause list
+                          || Function.of_arity(4)) = Exn.Fail,
+          // intended to be a single short line
+          msg :: ReadableString,
+          // if non-empty, ";" is added to the end of `message`, and
+          // each `detail` is shown on its own line with a 1-space prefix
+          ~details: [detail :: ReadableString, ...] = [],
+          // intended to be constructed by `error.val`, `error.annot`,
+          // or `error.text`; each `clause` starts its own line with
+          // a 2-space prefix:
+          clause :: error.Clause,
+          ...):
+  let msg = message(~srcloc: srcloc,
+                    ~who: who,
+                    ~realm: realm,
+                    msg,
+                    ~detail: [detail, ...],
+                    clause, ...)
+  let marks = Continuation.Marks.current()
+  let e:
+    match exn
+    | exn :: Function.of_arity(4): exn(msg, marks, [srcloc], [clause, ...])
+    | exn :: Function.of_arity(3): exn(msg, marks, [srcloc])
+    | exn :: Function.of_arity(2): exn(msg, marks)
+  throw e
+
+// exported as `error.message`, same arguments as `error` except for `~exn`
+fun message(~srcloc: srcloc :: maybe(Srcloc) = #false,
+            ~who: who :: maybe(error.Who) = #false,
+            ~realm: realm :: Symbol = #'rhombus,
+            message :: ReadableString,
+            ~detail: [detail :: ReadableString, ...] = [],
+            clause :: error.Clause,
+            ...) :~ String:
+   (if srcloc | Srcloc.to_report_string(srcloc) ++ ": " | "")
+   base.#{error-message->adjusted-string}(
+     (who && Symbol.from_string(to_string(who))),
```

Comment:
I’ve been wanting to ask this, but should we try to unify what is conventionally accepted as name argument at some point?  Now, `error` accepts a very lenient `ReadableString || Symbol || Identifier || Operator`, while some accept `ReadableString || Symbol` or just `Symbol`, and yet some accept just `String`.  Why can’t we just accept `String` and let user does their own conversion?  And why specifically `Identifier || Operator`, but not `Name`?

Reply:
I agree about unifying and about using `Name` if `Identifier || Operator` remains relevant. I still think it will be handy to be flexible here to avoid boilerplate conversions in common cases. Maybe we should allow any value for the name, since `to_string` should generally do the right thing?

Reply:
An annotation like `Any.to_string` (akin to `Any.to_boolean`) can be useful if we do allow any value.

Reply:
I remember now why `Any` (or `Any.to_string`) is probably not a good choice in this case: because `#false` s supposed to mean that nothing should be prefixed, not that `#false: ` should be a message prefix.